### PR TITLE
Pygalois

### PR DESCRIPTION
**Pull request by gurbinder533**
Merging current progress on Cython end. Bfs and Pagerank applications are working fine.

### CODE REVIEW

**Comment by ddn0** [comment]
- I'd leave SSSP out if it is not working at all.
  - Generally, I'd prefer to keep WIP out of any merged commits while POC can be publishable though preferably under a feature branch.
- I'd reword the commit messages to the traditional style: https://commit.style/


@insertinterestingnamehere this is POC of Cython integration that structurally is there but follow-on work needs to fill-in the actual implementation. I think it deserves a little more visibility either as a long-lived feature branch or feature flag either here or over at iss/Galois. Do you have a preference on how? And comments in the content of the PR also :D

File: python/galois/_bfs.pyx [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                loopname("not_visited_op"))
+
+    gPrint(b"Max distance : ", maxDist.reduce(), b"\n")
+#
+# Main callsite for Bfs
+#        
+def bfs(int numThreads, unsigned long source, string filename):
+    ## Hack: Need a better way to initialize shared
+    ## memory runtime.
+    sys = new SharedMemSys()
```
Comment (insertinterestingnamehere) [inline python/galois/_bfs.pyx:175]: You may have already tried this, but initializing SharedMemSys outside of a given function ought to work. It'll run at import time for the module you put it in. Given that it initializes a bunch of shared-object local data in Galois, I think it makes sense to just have a dedicated Python extension module that runs this initialization and then have the apps all import it. That all should work as long as galois itself is built as a shared object.

**Comment by insertinterestingnamehere** [comment]
I gave it a quick look-through and it looks reasonable. scikit-build is one good way to use Cython within CMake. WRT how to include this in Galois, I'd prefer something that can be enabled via a flag and not a separate branch, but that's not a strong preference. Whatever is more convenient for you.

**Comment by insertinterestingnamehere** [comment]
Follow-up note about initialization: If the only thing `SharedMemSys()` does is initialize data that's local to the Galois shared object, is there any reason we shouldn't just make that happen implicitly at load time? We really only gain the benefit of good resource management if we're *not* using static data at all.

**Comment by ddn0** [comment]
>  WRT how to include this in Galois, I'd prefer something that can be enabled via a flag and not a separate branch, but that's not a strong preference. Whatever is more convenient for you.

If you don't mind feature flagged stuff in the next Galois release, we can aim to merge this into iss/Galois then?

> Follow-up note about initialization: If the only thing SharedMemSys() does is initialize data that's local to the Galois shared object, is there any reason we shouldn't just make that happen implicitly at load time? We really only gain the benefit of good resource management if we're not using static data at all.

Module-level initialization is the eventual goal. There appears to some issue where default constructors weren't being called without `new` as well (more investigation needed), so the current state is just trying to get to some working state.

I've been bitten by Python-native code initialization issues so I think explicit resource management is nicer, but given the normal practice of Python code, we have to support implicit initialization for sure.

**Comment by insertinterestingnamehere** [comment]
> Module-level initialization is the eventual goal. There appears to some issue where default constructors weren't being called without `new` as well (more investigation needed), so the current state is just trying to get to some working state.
> 
> I've been bitten by Python-native code initialization issues so I think explicit resource management is nicer, but given the normal practice of Python code, we have to support implicit initialization for sure.

Clarification to my question here: independent of anything Python-related, why aren't we just initializing everything at shared object load-time in libgalois?

**Comment by insertinterestingnamehere** [comment]
Also, yes. Feature flagged stuff in the next Galois release seems fine to me.

**Comment by ddn0** [comment]
> Clarification to my question here: independent of anything Python-related, why aren't we just initializing everything at shared object load-time in libgalois?

Historically? I think we assumed everything was static so this didn't come up.

Practically? The memory pools are easier to track in bulk rather than implementing reference counting. And if you do that, the least we can do is provide some way for the user to reclaim all the memory Galois is hogging.
